library/pkcs12.c

```diff
@@ -209,6 +212,105 @@ static void pkcs12_fill_buffer( unsigned char *data, size_t data_len,
     }
 }
 
+
+static int calculate_hashes( mbedtls_md_type_t md_type, int iterations,
+    unsigned char *diversifier, unsigned char *salt_block,
+    unsigned char *pwd_block, unsigned char *hash_output, int use_salt,
+    int use_password, size_t hlen, size_t v )
+{
+#if defined(MBEDTLS_MD_C)
+    int ret = -1;
+    size_t i;
+    const mbedtls_md_info_t *md_info;
+    mbedtls_md_context_t md_ctx;
+    md_info = mbedtls_md_info_from_type( md_type );
+    if( md_info == NULL )
+        return ( MBEDTLS_ERR_PKCS12_FEATURE_UNAVAILABLE );
+
+    mbedtls_md_init( &md_ctx );
+
+    if( ( ret = mbedtls_md_setup( &md_ctx, md_info, 0 ) ) != 0 )
+        return ( ret );
+    // Calculate hash( diversifier || salt_block || pwd_block )
+    if( ( ret = mbedtls_md_starts( &md_ctx ) ) != 0 )
+        goto exit;
+
+    if( ( ret = mbedtls_md_update( &md_ctx, diversifier, v ) ) != 0 )
+        goto exit;
+
+    if( use_salt != 0 )
+    {
+        if( ( ret = mbedtls_md_update( &md_ctx, salt_block, v ) ) != 0 )
+            goto exit;
+    }
+
+    if( use_password != 0 )
+    {
+        if( ( ret = mbedtls_md_update( &md_ctx, pwd_block, v ) ) != 0 )
+            goto exit;
+    }
+
+    if( ( ret = mbedtls_md_finish( &md_ctx, hash_output ) ) != 0 )
+        goto exit;
+
+    // Perform remaining ( iterations - 1 ) recursive hash calculations
+    for( i = 1; i < (size_t) iterations; i++ )
+    {
+        if( ( ret = mbedtls_md( md_info, hash_output, hlen, hash_output ) )
+            != 0 )
+            goto exit;
+    }
+
+exit:
+    mbedtls_md_free( &md_ctx );
+    return ret;
+#else
+    psa_hash_operation_t op = PSA_HASH_OPERATION_INIT;
+    psa_algorithm_t alg = mbedtls_psa_translate_md( md_type );
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    size_t i, out_len, out_size = PSA_HASH_LENGTH( alg );
+
+    if( alg == PSA_ALG_NONE )
+        return ( MBEDTLS_ERR_PKCS12_FEATURE_UNAVAILABLE );
+
+    if( ( status = psa_hash_setup( &op, alg ) ) != PSA_SUCCESS )
+        goto exit;
+
+    // Calculate hash( diversifier || salt_block || pwd_block )
+    if( ( status = psa_hash_update( &op, diversifier, v ) ) != PSA_SUCCESS )
+        goto exit;
+
+    if( use_salt != 0 )
+    {
+        if( ( status = psa_hash_update( &op, salt_block, v ) ) != PSA_SUCCESS )
+            goto exit;
+    }
+
+    if( use_password != 0 )
+    {
+        if( ( status = psa_hash_update( &op, pwd_block, v ) ) != PSA_SUCCESS )
+            goto exit;
+    }
+
+    if( ( status = psa_hash_finish( &op, hash_output, out_size, &out_len ) )
+        != PSA_SUCCESS )
+        goto exit;
+
+    // Perform remaining ( iterations - 1 ) recursive hash calculations
+    for( i = 1; i < (size_t) iterations; i++ )
+    {
+        if( ( status = psa_hash_compute( alg, hash_output, hlen, hash_output,
+            out_size, &out_len ) ) != PSA_SUCCESS )
+            goto exit;
+    }
+
+exit:
+    psa_hash_abort( &op );
```

Comment:
Result of `psa_hash_abort()` is ignored. I suggest to also check the status of `psa_hash_abort()`.

Reply:
I'll wait for any other review comments and fix this one with others in bulk.